File: ps/middle-of-the-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        ListNode* slow = head;
        ListNode* fast = head;
        while (fast->next != NULL) {
            slow = slow->next;
            fast = fast->next;
            if (fast->next != NULL) fast = fast->next;
        }
        return slow;
    }
};

File: ps/middle-of-the-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        nodeI = head
        nodeJ = head
        while nodeJ != None:
            nodeJ = nodeJ.next
            if nodeJ == None:
                return nodeI
            nodeI = nodeI.next
            nodeJ = nodeJ.next
        return nodeI


